File: leetcode/663-equal-tree-partitiion/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


"""
    questions to ask:
    - negative numbers?
"""


class Solution(object):
    def checkEqualTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool

        1st approach: dfs + hashtable
        - for each node, calculate its children sum
        - calculate the total sum as well
        - dfs again to check if total - cur == cur

        corner cases:
        - [0,null,0]
        - [0,-1,1]

        88 ms, faster than 21.48%
        """
        m = {}
        total = self.calSum(root, m)

        stack = [root]
        while len(stack) > 0:
            pop = stack.pop()

            # check if total is in the map
            # important: we can only remove edge(not root)
            if pop != root and total == 2*m[pop]:
                return True

            if pop.left != None:
                stack.append(pop.left)
            if pop.right != None:
                stack.append(pop.right)
        return False

    def calSum(self, node, m):
        if node == None:
            return 0
        left = self.calSum(node.left, m)
        right = self.calSum(node.right, m)
        curSum = left + node.val + right
        m[node] = curSum
        return curSum


class Solution(object):
    def checkEqualTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool

        2nd approach: dfs + hashset, learned from others
        - for each node, calculate its children sum
        - calculate the total sum as well
        - dfs again to check if total - cur == cur

        compare to the 1st approach:
        - actually we dont need to care if there are duplicate sum values,
        because if there are, lets say 2 same values , it means that we can cut the tree in 2 ways

        ref:
        - https://leetcode.com/problems/equal-tree-partition/discuss/106727/javac-simple-solution-with-only-one-hashmap

        corner cases:
        - [0,null,0]
        - [0,-1,1]

        76 ms, faster than 45.64%
        """
        if root == None:
            return False
        m = set()
        # we should not add total to the hashset
        # otherwise we are not partitioning the tree
        total = root.val + \
            self.calSum(root.left, m) + self.calSum(root.right, m)
        return total % 2 == 0 and total / 2 in m

    def calSum(self, node, m):
        if node == None:
            return 0
        left = self.calSum(node.left, m)
        right = self.calSum(node.right, m)
        curSum = left + node.val + right
        m.add(curSum)
        return curSum

File: leetcode/663-equal-tree-partitiion/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	1st approach: dfs + hashtable
	- for each node, calculate its children sum
	- calculate the total sum as well
	- dfs again to check if total - cur == cur

	corner cases:
	- [0,null,0]
	- [0,-1,1]

	20 ms, faster than 83.33%
*/
func checkEqualTree(root *TreeNode) bool {
	if root == nil {
		return false
	}
	m := make(map[*TreeNode]int)
	total := calSum(root, m)

	stack := []*TreeNode{root}
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if pop != root && total == 2*m[pop] {
			return true
		}
		if pop.Left != nil {
			stack = append(stack, pop.Left)
		}
		if pop.Right != nil {
			stack = append(stack, pop.Right)
		}
	}
	return false
}

func calSum(node *TreeNode, m map[*TreeNode]int) int {
	if node == nil {
		return 0
	}
	left := calSum(node.Left, m)
	right := calSum(node.Right, m)
	sum := left + node.Val + right
	m[node] = sum
	return sum
}

/*
	2nd approach: dfs + hashset, learned from others
	- for each node, calculate its children sum
	- calculate the total sum as well
	- dfs again to check if total - cur == cur

	compare to the 1st approach:
	- actually we dont need to care if there are duplicate sum values,
	because if there are, lets say 2 same values , it means that we can cut the tree in 2 ways

	ref:
	- https://leetcode.com/problems/equal-tree-partition/discuss/106727/javac-simple-solution-with-only-one-hashmap

	corner cases:
	- [0,null,0]
	- [0,-1,1]

	32 ms, faster than 16.67%
*/
func checkEqualTree1(root *TreeNode) bool {
	if root == nil {
		return false
	}
	m := make(map[int]bool)
	total := root.Val + calSum1(root.Left, m) + calSum1(root.Right, m)
	// check if total is even
	if total%2 != 0 {
		return false
	}
	if _, x := m[total/2]; x {
		return true
	}
	return false
}

func calSum1(node *TreeNode, m map[int]bool) int {
	if node == nil {
		return 0
	}
	left := calSum1(node.Left, m)
	right := calSum1(node.Right, m)
	sum := left + node.Val + right
	m[sum] = true
	return sum
}

func main() {

}
